middle_end/flambda2/types/provers.ml

```diff
@@ -630,7 +630,10 @@ let prove_is_array_with_element_kind env t ~element_kind : _ proof =
   | Value Bottom -> Invalid
   | Value (Ok (Array { element_kind = Unknown; _ })) -> Unknown
   | Value (Ok (Array { element_kind = Known element_kind'; _ })) ->
-    Proved (K.With_subkind.equal element_kind' element_kind)
+    Proved
+      (K.equal
```

Comment:
Maybe here, you could instead make use of the `K.With_subkind.compatible` function ?

Reply:
I started with that, but I'm worried that `compatible` is not symmetrical.
We want to support both cases where the primitive is specialised and the type isn't (like in the example), and the more common case where the primitive is not specialised but the type is.

Reply:
Both of these cases could be invalid, though, no? `Array_load (Immediates, _)` on a value array and `Array_set (Values, _)` on an immediate array could both be bad.

Reply:
`Array_load (Immediates, _)` on a value array returns an immediate (it merely acts as an additional type annotation, just like you could refine a generic type into a specific one using GADTs).
`Array_set (Values, _)` on an immediate array works in the same way: it merely means that in the current context, the primitive can be specialised.

Here are two examples showing those cases:
```ocaml
type _ w = Int : int w | Other : _ w
let load_immediate_on_value_array (type a) (wit : a w) (v : a) =
  let make_array x = [|x|] (* here we get a value array *) in
  let get_int (a : int array) = a.(0) (* here we load with kind Immediates *) in
  let a = make_array v in
  match wit with
  | Int -> get_int a
  | Other -> 0

let set_value_on_immediate_array () =
  let set (a : _ array) x = a.(0) <- x (* Value set *) in
  let a = [|0|] (* Immediate array *) in
  set a 0
```

The example from this PR shows the first case but without GADTs.